pkg/sidecar/sidecar.go

```diff
@@ -17,59 +17,96 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/andres-erbsen/clock"
+	proto "github.com/spiffe/go-spiffe/proto/spiffe/workload"
 	"github.com/spiffe/spire/api/workload"
-	proto "github.com/spiffe/spire/proto/api/workload"
+	"golang.org/x/sys/unix"
 )
 
-// sidecar is the component that consumes the Workload API and renews certs
+// Config contains config variables when creating a SPIFFE Sidecar.
+type Config struct {
+	AgentAddress       string
+	Cmd                string
+	CmdArgs            string
+	CertDir            string
+	SvidFileName       string
+	SvidKeyFileName    string
+	SvidBundleFileName string
+	RenewSignal        string
+	Timeout            string
+	ExternalProcess    bool
+	ProcessReloadFunc  func() error
+}
+
+// Sidecar is the component that consumes the Workload API and renews certs
 // implements the interface Sidecar
-type sidecar struct {
-	config            *SidecarConfig
+type Sidecar struct {
+	config            *Config
 	processRunning    int32
 	process           *os.Process
 	workloadAPIClient workload.X509Client
+	certReadyChan     chan struct{}
 }
 
 const (
 	// default timeout Duration for the workloadAPI client when the defaultTimeout
 	// is not configured in the .conf file
-	defaultTimeout = time.Duration(5 * time.Second)
+	defaultTimeout = 5 * time.Second
+	delayMin       = time.Second
+	delayMax       = time.Minute
 
 	certsFileMode = os.FileMode(0644)
 	keyFileMode   = os.FileMode(0600)
 )
 
-// NewSidecar creates a new sidecar
-func NewSidecar(config *SidecarConfig) (*sidecar, error) {
+// NewSidecar creates a new SPIFFE sidecar
+func NewSidecar(config *Config) (*Sidecar, error) {
 	timeout, err := getTimeout(config)
 	if err != nil {
 		return nil, err
 	}
 
-	return &sidecar{
+	return &Sidecar{
 		config:            config,
 		workloadAPIClient: newWorkloadAPIClient(config.AgentAddress, timeout),
+		certReadyChan:     make(chan struct{}, 1),
 	}, nil
 }
 
 // RunDaemon starts the main loop
 // Starts the workload API client to listen for new SVID updates
 // When a new SVID is received on the updateChan, the SVID certificates
 // are stored in disk and a restart signal is sent to the proxy's process
-func (s *sidecar) RunDaemon(ctx context.Context) error {
+func (s *Sidecar) RunDaemon(ctx context.Context) {
 	// Create channel for interrupt signal
 	interrupt := make(chan os.Signal, 1)
```

Comment:
TERM and INT should probably not be handled from inside the library since each application would have its own requirements on what causes `RunDaemon` to return. I'd suggest moving the signal handling back out into `main` and change it to just cancel the context that is passed into `RunDaemon`.

Reply:
Done.